File: YukariLauncher/src/main/java/net/kdt/pojavlaunch/tasks/AsyncMinecraftDownloader.java
package net.kdt.pojavlaunch.tasks;

import com.arata.yukarilauncher.event.sticky.MinecraftVersionValueEvent;

import net.kdt.pojavlaunch.JMinecraftVersionList;

import org.greenrobot.eventbus.EventBus;

/**
 * マインクラフトのバージョン情報を非同期で取得するためのユーティリティクラス。
 */
public class AsyncMinecraftDownloader {

    /**
     * バージョン文字列からバージョン情報を検索します。
     * @param versionString 検索するバージョンID
     * @return 見つかったバージョン、存在しない場合はnull
     */
    public static JMinecraftVersionList.Version getListedVersion(String versionString) {
        JMinecraftVersionList versionList = getJMinecraftVersionList();
        if (versionList == null || versionList.versions == null) return null;
        for (JMinecraftVersionList.Version version : versionList.versions) {
            if (version.id.equals(versionString)) return version;
        }
        return null;
    }

    /**
     * イベントバスからJMinecraftVersionListを取得します。
     */
    private static JMinecraftVersionList getJMinecraftVersionList() {
        MinecraftVersionValueEvent event = EventBus.getDefault().getStickyEvent(MinecraftVersionValueEvent.class);
        if (event != null) return event.getList();
        else return null;
    }

    /**
     * ダウンロード完了リスナーインターフェース。
     */
    public interface DoneListener{
        /** ダウンロードが完了したときに呼び出されます。 */
        void onDownloadDone();
        /** ダウンロードが失敗したときに呼び出されます。 */
        void onDownloadFailed(Throwable throwable);
    }
}

File: YukariLauncher/src/main/java/com/arata/yukarilauncher/task/AsyncMinecraftDownloader.kt
package com.arata.yukarilauncher.task

import com.arata.yukarilauncher.event.sticky.MinecraftVersionValueEvent
import com.arata.yukarilauncher.value.JMinecraftVersionList
import org.greenrobot.eventbus.EventBus

/** 非同期Minecraftダウンローダー。 */
object AsyncMinecraftDownloader {

    /** バージョン文字列からリスト内のバージョンを取得する。 @param versionString バージョン文字列 @return バージョン情報、見つからない場合はnull */
    fun getListedVersion(versionString: String): JMinecraftVersionList.Version? {
        val versionList = getJMinecraftVersionList() ?: return null
        for (version in versionList.versions!!) {
            if (version.id == versionString) return version
        }
        return null
    }

    /** JMinecraftVersionListをEventBusから取得する。 @return バージョンリスト */
    private fun getJMinecraftVersionList(): JMinecraftVersionList? {
        val event = EventBus.getDefault().getStickyEvent(MinecraftVersionValueEvent::class.java)
        return event?.list
    }

    /** ダウンロード完了リスナー。 */
    interface DoneListener {
        /** ダウンロード完了時に呼び出される。 */
        fun onDownloadDone()
        /** ダウンロード失敗時に呼び出される。 @param throwable 例外 */
        fun onDownloadFailed(throwable: Throwable)
    }
}